src/socket.cpp

```diff
@@ -139,50 +142,58 @@ void Socket::send(
 	const auto sent  = low_write(
 	  ::sendto, sockfd, _msg, flags, (sockaddr *) &_addr, sizeof(_addr));
 
-	if (sent == -1) {
-		// TODO: Check for sent == EAGAIN or EWOULDBLOCK and
-		// don't throw exception in that case.
-		throw std::runtime_error(net::method::getErrorMsg());
+	const auto currentErrNo = errno;
+	errno                   = 0;
+	if (sent == -1 && currentErrNo != EAGAIN && currentErrNo != EWOULDBLOCK) {
+		throw std::runtime_error(net::method::getErrorMsg(currentErrNo));
 	}
 }
 
 
-std::string Socket::read(const int _bufSize) const
+std::string Socket::read(const int _bufSize, bool *_isSpecialErrorInNonBlocking) const
 {
 	std::string str;
 	str.reserve(_bufSize);
 
 	const auto recvd = low_read(::read, sockfd, str);
 
+	const auto currentErrNo = errno;
+	errno                   = 0;
 	if (recvd == -1) {
-		// TODO: Check for bytes == EAGAIN or EWOULDBLOCK and
-		// don't throw exception in that case.
-		throw std::runtime_error(net::method::getErrorMsg());
+		if ( !_isSpecialErrorInNonBlocking || (_isSpecialErrorInNonBlocking && currentErrNo != EAGAIN && currentErrNo != EWOULDBLOCK))	// blocking & non blocking failure case
+			throw std::runtime_error(net::method::getErrorMsg(currentErrNo));
+		// else if ( _isSpecialErrorInNonBlocking && currentErrNo != EAGAIN && currentErrNo != EWOULDBLOCK) // non blocking failure case
```

Comment:
Does not throws [`invalid_argument`](http://en.cppreference.com/w/cpp/error/invalid_argument) in case `nullptr` is sent/or last param is ignored and non-blocking error occurs.